### ISSUE TITLE
[net6] decompressing CMS content randomly fails

### ISSUE THREAD

**Issue by jstedfast** [bcgit/bc-csharp#423]
**BouncyCastle.Cryptography:** v2.1.1 (tried NuGet packages and source)
**OS Platform:** Windows (haven't tested on other OS's yet)
**Framework:** net6.0
**Exception:** System.IO.InvalidDataException : The archive entry was compressed using an unsupported compression method.
**Stack Trace:**
```
Inflater.Inflate(FlushCode flushCode)
Inflater.ReadInflateOutput(Byte* bufPtr, Int32 length, FlushCode flushCode, Int32& bytesRead)
Inflater.ReadOutput(Byte* bufPtr, Int32 length, Int32& bytesRead)
Inflater.InflateVerified(Byte* bufPtr, Int32 length)
CopyToStream.Write(Byte[] buffer, Int32 offset, Int32 count)
Stream.CopyTo(Stream destination, Int32 bufferSize)
CopyToStream.CopyFromSourceToDestination()
DeflateStream.CopyTo(Stream destination, Int32 bufferSize)
ZLibStream.CopyTo(Stream destination, Int32 bufferSize)
FilterStream.CopyTo(Stream destination, Int32 bufferSize) line 30
SecureMimeContext.DecompressTo(Stream stream, Stream output, CancellationToken cancellationToken) line 624
SecureMimeTestsBase.TestSecureMimeCompressionWithContext() line 415
```

I'm currently working on porting MimeKit to BouncyCastle 2.x and I've run into a number of problems with my PGP and S/MIME unit tests failing (so expect more of these bug reports as I dig into them a bit more).

One of my unit tests randomly fails. As you can see from the stacktrace above, it fails inside of `DecompressTo()` which I will paste below.

I took a look at `FilterStream.CopyTo()` and noticed that it overrides the base class implementation in order to call `s.CopyTo(destination, bufferSize)`, but only for netstandard2.1+ (which also affects net6.0).

Out of curiosity, I added my own `CopyTo()` method (included with the source snippet below) and if I use that instead of `content.ContentStream.CopyTo (output, 4096);`, then I no longer get failures.

```csharp
static void CopyTo (Stream input, Stream output, int bufferSize)
{
	var buffer = ArrayPool<byte>.Shared.Rent (bufferSize);

	try {
		int nread;

		while ((nread = input.Read (buffer, 0, bufferSize)) > 0) {
			output.Write (buffer, 0, nread);
		}
	} finally {
		ArrayPool<byte>.Shared.Return (buffer);
	}
}

public virtual void DecompressTo (Stream stream, Stream output, CancellationToken cancellationToken = default (CancellationToken))
{
	if (stream == null)
		throw new ArgumentNullException (nameof (stream));

	if (output == null)
		throw new ArgumentNullException (nameof (output));

	cancellationToken.ThrowIfCancellationRequested ();

	using (var parser = new CmsCompressedDataParser (stream)) {
		var content = parser.GetContent ();

		try {
			cancellationToken.ThrowIfCancellationRequested ();
			content.ContentStream.CopyTo (output, 4096);
			//CopyTo (content.ContentStream, output, 4096);
		} finally {
			content.ContentStream.Dispose ();
		}
	}
}
```

**Comment by jstedfast** [comment]
I just modified my UnitTests project to be net48 instead of negt6.0 and now all my BouncyCastle-related tests pass.

It looks like there are likely bugs in the conditional codepaths that affect net6.0 (like the Stream.CopyTo override and maybe some of the Span<byte> code-paths).

**Comment by peterdettman** [comment]
Thanks, happy to see you migrating. I've forked MimeKit so I can help out.

There's another difference in net6.0 which is that we use ZLibStream there (and our own implementation otherwise). I've done some experimenting on the compressed data failure (e.g. completely removing FullReaderStream from CmsTypedStream, so then CopyTo is failing even on a BufferedStream(ZLibStream) i.e. not even our code). I am suspecting some sort of issue around Dispose/leaveOpen, but still looking into it.

**Comment by jstedfast** [comment]
@peterdettman 

Thanks! I was going to try and work on writing more isolated tests for you since I know you are a busy guy, but I appreciate it!

I have a branch setup for the BouncyCastle 2.x port called `bouncycastle2.0`. I also have a local change that I can commit (if that would help you) that switches over to using a side-by-side checkout of bc-csharp from source as opposed to using the NuGet package.

That's how I discovered some of the things I was doing wrong.

> There's another difference in net6.0 which is that we use ZLibStream there (and our own implementation otherwise).

Yea, I noticed that - that's how I also noticed the CopyTo() override. I had no idea that was even overridable - haha.

There's clearly no bug in the FilterStream.CopyTo() override, but as I mentioned in my bug report, using my own implementation of CopyTo() which just uses Read/Write, that somehow fixes the ZLibCompression bug but I wasn't able to figure out why yet.

> I am suspecting some sort of issue around Dispose/leaveOpen

Yea, my thought was either that -or- something to do with the Read/Write overrides that take Spans instead of byte arrays. I figured that the dotnet Stream implementation of CopyTo() was probably taking advantage of those APIs existing and making use of them and therefore, bypassing them and using a Read/Write loop exposed that kind of bug.

Obviously I'm just guessing at this point.

**Comment by jstedfast** [comment]
Oh, and before I forget... I think it might be useful to have a `leaveOpen` argument on all of the `Cms*Parser` .ctors if they dispose that input stream. I don't think any of my unit tests are re-using said input stream (that was one of my first suspects), but they *do* dispose of it themselves.

**Comment by jstedfast** [comment]
@peterdettman Thanks for looking into this. You might be on to something here because I just found a re-use (and a double Return() of the same buffer) in the PGP Encrypt() logic. Looks like I did a refactoring of the code in order to re-use the same buffer but didn't fix the code that returned it to the pool (and then used it after the first Return()).

I'll commit my fix to MimeKit master and rebase the bouncycastle2.0 branch on it to see if that resolves the test failures. I suspect it will.
